performance.md

```diff
@@ -0,0 +1,341 @@
+takes too much time.
+
+### Checking Assembly Code
+
+The function [`disassemble`][disassemble] takes a function and returns the
+machine code of it. For example:
+
+~~~lisp
+* (defun plus (a b)
+    (+ a b))
+PLUS
+
+* (disassemble 'plus)
+; disassembly for PLUS
+; Size: 37 bytes. Origin: #x52B8063B
+; 3B:       498B5D60         MOV RBX, [R13+96]                ; no-arg-parsing entry point
+                                                              ; thread.binding-stack-pointer
+; 3F:       48895DF8         MOV [RBP-8], RBX
+; 43:       498BD0           MOV RDX, R8
+; 46:       488BFE           MOV RDI, RSI
+; 49:       FF1425B0001052   CALL QWORD PTR [#x521000B0]      ; GENERIC-+
+; 50:       488B75E8         MOV RSI, [RBP-24]
+; 54:       4C8B45F0         MOV R8, [RBP-16]
+; 58:       488BE5           MOV RSP, RBP
+; 5B:       F8               CLC
+; 5C:       5D               POP RBP
+; 5D:       C3               RET
+; 5E:       CC0F             BREAK 15                         ; Invalid argument count trap
+~~~
+
+## Using Declare Expression
+
+The [*declare expression*][declare] can be used to provide hints for compilers
+to perform various optimization. Please note that these hints are
+implementation-dependent. Some implementations such as SBCL support this
+feature, and you may refer to their own documents for detailed
+information. Here only some basic techniques mentioned in CLHS are introduced.
+
+In general, declare expressions can occur only at the beginning of the bodies
+of certain forms, or immediately after a documentation string if the context
+allows. Also, the content of a declare expression is restricted to limited
+forms. Here we introduce some of them that are related to performance tuning.
+
+Please keep in mind that these optimization skills introduced in this section
+are strongly connected to the Lisp implementation selected. Always check their
+documents before using `declare`!
+
+### Speed and Safety
+
+Lisp allows you to specify several quality properties for the compiler using
+the declaration [`optimize`][optimize]. Each quality may be assigned a value
+from 0 to 3, with 0 being "totally unimportant" and 3 being "extremely
+important".
+
+The most significant qualities might be `safety` and `speed`. 
+
+By default, Lisp considers code safety to be much more important than
+speed. But you may adjust the weight for more aggressive optimization.
+
+~~~lisp
+* (defun max-original (a b)
+    (max a b))
+MAX-ORIGINAL
+
+* (disassemble 'max-original)
+; disassembly for MAX-ORIGINAL
+; Size: 144 bytes. Origin: #x52D450EF
+...
+
+* (defun max-with-speed-3 (a b)
+    (declare (optimize (speed 3) (safety 0)))
+    (max a b))
+MAX-WITH-SPEED-3
+
+* (disassemble 'max-with-speed-3)
+; disassembly for MAX-WITH-SPEED-3
+; Size: 92 bytes. Origin: #x52D452C3
+...
+~~~
+
+As you can see, the generated assembly code is much shorter, which means that
+unused code are pruned.
```

Comment:
I'm not an expert, but I suspect the shorter code is omitting type checks and other safety features when `(safety 0)` is set. I would hope the compiler prunes unreachable code in all cases.

Reply:
Well, "pruning all unreachable code" is hard.

SBCL can give you some warnings about unused variables or so, but this is the best it can do.

If I recall correctly, LispWorks has function to remove unused code to while dumping executable, so it may produce small executable.

SBCL cannot do that for sure.